dbms/src/Storages/S3/FileCache.cpp

```diff
@@ -0,0 +1,552 @@
+        return nullptr;
+    }
+
+    auto file_seg = std::make_shared<FileSegment>(toLocalFilename(s3_key), FileSegment::Status::Empty, estimzted_size, file_type);
+    table.set(s3_key, file_seg);
+    bgDownload(s3_key, file_seg);
+
+    return nullptr;
+}
+
+// Remove `local_fname` from disk and remove parent directory if parent directory is empty.
+void FileCache::removeDiskFile(const String & local_fname)
+{
+    try
+    {
+        for (std::filesystem::path p(local_fname); p.is_absolute() && std::filesystem::exists(p); p = p.parent_path())
+        {
+            auto s = p.string();
+            if (s != cache_dir && (s == local_fname || std::filesystem::is_empty(p)))
+            {
+                std::filesystem::remove(p);
+            }
+            else
+            {
+                break;
+            }
+        }
+    }
+    catch (std::exception & e)
+    {
+        LOG_WARNING(log, "Throw exception in removeFile: ", e.what());
+    }
+    catch (...)
+    {
+        tryLogCurrentException("Throw exception in removeFile");
+    }
+}
+
+void FileCache::remove(const String & s3_key)
+{
+    auto file_type = getFileType(s3_key);
+    auto & table = tables[static_cast<UInt64>(file_type)];
+
+    std::lock_guard lock(mtx);
+    auto f = table.get(s3_key);
+    if (f == nullptr)
+    {
+        return;
+    }
+    std::ignore = removeImpl(table, s3_key, f);
+}
+
+std::pair<UInt64, std::list<String>::iterator> FileCache::removeImpl(LRUFileTable & table, const String & s3_key, FileSegmentPtr & f)
+{
+    // Except currenly thread and the FileTable,
+    // there are other threads hold this FileSegment object.
+    if (f.use_count() > 2)
+    {
+        return {0, {}};
+    }
+    ProfileEvents::increment(ProfileEvents::FileCacheEvict);
+    const auto & local_fname = f->getLocalFileName();
+    removeDiskFile(local_fname);
+    auto temp_fname = toTemporaryFilename(local_fname);
+    removeDiskFile(temp_fname);
+
+    auto release_size = f->getSize();
+    releaseSpaceImpl(release_size);
+    return {release_size, table.remove(s3_key)};
+}
+
+bool FileCache::reserveSpaceImpl(FileType reserve_for, UInt64 size, bool try_evict)
+{
+    if (cache_used + size <= cache_capacity)
+    {
+        cache_used += size;
+        return true;
+    }
+    if (try_evict)
+    {
+        UInt64 min_evict_size = size - (cache_capacity - cache_used);
+        LOG_DEBUG(log, "tryEvictFile for {} min_evict_size={}", magic_enum::enum_name(reserve_for), min_evict_size);
+        tryEvictFile(reserve_for, min_evict_size);
+        return reserveSpaceImpl(reserve_for, size, /*try_evict*/ false);
+    }
+    return false;
+}
+
+// The basic evict logic:
+// 1. Distinguish cache priority according to file type. The larger the file type, the lower the priority.
+// 2. Files with higher priority can be evicted from files with lower priority.
+// 3. Files of the same priority can only evict files that have not been accessed recently.
```

Comment:
Looks like this algorithm will leave many useless files with small `FileType` cannot get evicted. Because we always evict the files with largest `FileType`